### PR TITLE
cmake: kconfig: Add new abstraction for -Oz optimization level

### PR DESCRIPTION
**Pull request by jonathonpenix**
Both Clang [1] and (recently) GCC [2] support this flag to enable additional codesize optimizations beyond -Os, possibly at the expense of performance.

This tradeoff is worthwhile for some (and, Clang's -Oz seems to be closer to GCC's -Os currently), so add a new abstraction for this flag so users can select it as appropriate.

Additionally, clang's -Oz seems to be closer to GCC's -Os (clang's -Os seems to be much more agressive in terms of inlining, etc.), so set the default opt level for clang/llvm to be -Oz for the better codesize.

[1] https://clang.llvm.org/docs/CommandGuide/clang.html#cmdoption-O0
[2] https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#index-Oz

### CODE REVIEW

File: Kconfig.zephyr [hunk 1]
```diff
@@ -490,6 +491,12 @@ config SIZE_OPTIMIZATIONS
 	  Compiler optimizations will be set to -Os independently of other
 	  options.
 
+config MIN_SIZE_OPTIMIZATIONS
```
Comment (jonathonpenix) [inline Kconfig.zephyr:494]: Please let me know if anyone has better naming suggestions here--I don't love "min size optimizations" but I'm struggling to come up with a better name.

**Review by jonathonpenix — COMMENTED** [review]
Please note that I did **not** add a new Picolibc-specific option here: https://github.com/zephyrproject-rtos/zephyr/blob/3570408db556fbb53cd775ae3d2e44553e28cb31/lib/libc/picolibc/Kconfig#L73-L108

As best I can tell (from basic grepping around and setting these while building picolibc as a module), these options are unused/have no effect. @keith-packard (or anyone else knowledgeable in this area!) could you please confirm if these are unused or if there are any plans to use these in the future? Apologies in advance if I missed something and I can add a new Picolibc-specific option for -Oz if needed!

That said, I did create a "companion" PR in picolibc to set `PREFER_SIZE_OVER_SPEED` in picolibc based on the new `CONFIG_MIN_SIZE_OPTIMIZATIONS` option added here. Please see https://github.com/picolibc/picolibc/pull/737

**Review by jeremybettis — DISMISSED** [review]
Love it, and it obsoletes my #70446

**Review by aescolar — DISMISSED** [review]
+1

**Review by ithinuel — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Review by nordicjm — DISMISSED** [review]

File: Kconfig.zephyr (hunk 1 above)
Comment (kartben) [reply Kconfig.zephyr:494]: I would use the same prefix as the existing `SIZE_OPTIMIZATIONS` Kconfig, and use the same "aggressive" terminology from the compile flag, i.e.: `SIZE_OPTIMIZATIONS_AGGRESSIVE`

File: cmake/compiler/compiler_flags_template.cmake [hunk 2]
```diff
@@ -13,6 +13,8 @@ set_compiler_property(PROPERTY optimization_speed)
 
 set_compiler_property(PROPERTY optimization_size)
 
+set_compiler_property(PROPERTY optimization_min_size)
```
Comment (kartben) [inline cmake/compiler/compiler_flags_template.cmake:16]: nit: same comment regarding maybe using the same `optimization_size` prefix here (and therefore calling this `optimization_size_aggressive`)?

Comment (jonathonpenix) [reply cmake/compiler/compiler_flags_template.cmake]: Done!

File: Kconfig.zephyr (hunk 1 above)
Comment (jonathonpenix) [reply Kconfig.zephyr]: I like this, thanks for the suggestion! Made this change in the most recent version.

**Review by kartben — APPROVED** [review]

**Review by ithinuel — APPROVED** [review]

**Comment by kartben** [comment]
@tejlmand please help make sure this makes it into 3.7, if possible :)

**Review by nordicjm — APPROVED** [review]

**Comment by aescolar** [comment]
@tejlmand do you want to bless this PR?

**Review by dleach02 — APPROVED** [review]

**Pull request merged**